File: packages/sdk-js/core/ed25519/node/verify.js
/**
 * PEAC Protocol v0.9.1
 * Ed25519 signature verifier (Node.js)
 * Uses libsodium-wrappers.
 * Apache 2.0 License
 */

const sodium = require('libsodium-wrappers');

async function verify(message, signatureBase64, publicKeyBase64, nonce, timestamp) {
  await sodium.ready;
  const publicKey = Buffer.from(publicKeyBase64, 'base64');
  const signature = Buffer.from(signatureBase64, 'base64');
  const combined = Buffer.concat([
    Buffer.from(message),
    Buffer.from(nonce),
    Buffer.from(String(timestamp))
  ]);
  return sodium.crypto_sign_verify_detached(signature, combined, publicKey);
}

module.exports = { verify };

File: archived/core/ed25519/python/verify.py
"""
PEAC Protocol v0.9.1
Ed25519 signature verifier (Python)
Apache 2.0 License
"""

import base64
from nacl.signing import VerifyKey
from nacl.exceptions import BadSignatureError

def verify(message: str, signature_b64: str, public_key_b64: str, nonce: str, timestamp: int) -> bool:
    vk = VerifyKey(base64.b64decode(public_key_b64))
    combined = (message + nonce + str(timestamp)).encode('utf-8')
    try:
        vk.verify(combined, base64.b64decode(signature_b64))
        return True
    except BadSignatureError:
        return False
